otherlibs/unix/envir.c

```diff
@@ -13,18 +13,48 @@
 /*                                                                        */
 /**************************************************************************/
 
+#include <caml/config.h>
+
+#ifdef HAS_UNISTD
+#include <unistd.h>
+#endif
+#include <sys/types.h>
+
 #include <caml/mlvalues.h>
 #include <caml/alloc.h>
 
-#ifndef _WIN32
 extern char ** environ;
-#endif
 
-CAMLprim value unix_environment(value unit)
+CAMLprim value unix_environment_unsafe(value unit)
 {
   if (environ != NULL) {
     return caml_copy_string_array((const char**)environ);
   } else {
     return Atom(0);
   }
 }
+
+static char **secure_environ(void)
+{
+#if defined(HAS_ISSETUGID)
```

Comment:
Is this good enough? Under Linux you have `secure_getenv` but not `issetugid`. Surely you should use `getauxval` when available, as described in https://github.com/ocaml/ocaml/commit/38e2cd6a580e5b14a503f34d5ca7709d190c36a3#commitcomment-22736369 ?

Otherwise, your `environment` primitive is less secure than `getenv` under Linux.